### PR TITLE
Add torch.from_file function similar to the Storage.from_file, but returning tensor

### PR DESCRIPTION
**Pull request by VitalyFedyunin**
Porting `torch.Storage.from_file(filename, shared, size)` function to `torch.from_file(filename, shared, size, dtype=torch.int)`

### CODE REVIEW

File: test/test_torch.py [hunk 1]
```diff
@@ -9526,6 +9526,28 @@ def test_from_file(self):
             t2.fill_(rnum)
             self.assertEqual(t1, t2, 0)
 
+    # @unittest.skipIf(IS_WINDOWS, "TODO: need to fix this test case for Windows")
```
Comment (ezyang) [inline test/test_torch.py]: Delete comment?

File: aten/src/ATen/native/TensorFactories.cpp [hunk 2]
```diff
@@ -734,5 +736,34 @@ Tensor tensor_cuda(ArrayRef<T> values, const TensorOptions& options) {
   }
 AT_FORALL_SCALAR_TYPES_EXCEPT_HALF(TENSOR)
 #undef TENSOR
+
+Tensor from_file(std::string filename, c10::optional<bool> shared, c10::optional<long> size, const TensorOptions& options) {
+    AT_CHECK(!options.pinned_memory(), "tensors constructed from a file cannot be pinned");
+    size_t my_size = 0;
```
Comment (ezyang) [inline aten/src/ATen/native/TensorFactories.cpp]: Oh interesting. Is `size` really optional in this case? It looks to me that `size = 0` and `size = nullopt` are indistinguishable.

Actually, what I half expected to see was that the `nullopt` case would infer the number of elements based on the existing file size.

File: aten/src/ATen/native/TensorFactories.cpp [hunk 3]
```diff
@@ -734,5 +736,34 @@ Tensor tensor_cuda(ArrayRef<T> values, const TensorOptions& options) {
...
+    if (size.has_value())
+    {
+      my_size = size.value();
+    }
+    int flags = 0;
+    if (shared.has_value() && shared.value()) {
```
Comment (ezyang) [inline aten/src/ATen/native/TensorFactories.cpp]: Same with shared. Isn't this just a bool that defaults to `false`?

File: aten/src/ATen/native/TensorFactories.cpp (hunk 2 above)
Comment (ezyang) [reply aten/src/ATen/native/TensorFactories.cpp]: Oh, I see that you do set the size to `actual_size` if you do this. But you're testing if `my_size = 0` which is fishy; if I said `size = 0`, surely I really meant to give me a zero size view on the file? The docs don't disambiguate this case clearly, but a strict reading suggests that you should end up with a zero element tensor.

File: aten/src/ATen/native/TensorFactories.cpp [hunk 4]
```diff
@@ -734,5 +736,34 @@ Tensor tensor_cuda(ArrayRef<T> values, const TensorOptions& options) {
...
+      flags = TH_ALLOCATOR_MAPPED_SHARED;
+    }
+    size_t actual_size = -1;
+    auto dtype = options.dtype();
```
Comment (ezyang) [inline aten/src/ATen/native/TensorFactories.cpp]: You're completely ignoring the rest of options. It's probably better to just make this explicitly take only `DataType` and dispense with all the other arguments. Saves you an assert too.

**Review by ezyang — APPROVED** [review]
LGTM, but with comments

File: aten/src/ATen/native/TensorFactories.cpp (hunk 4 above)
Comment (VitalyFedyunin) [reply aten/src/ATen/native/TensorFactories.cpp]: Better, but our code generation currently out of `from_file(str filename, bool? shared=False, int? size=None, *, ScalarType? dtype=None) -> Tensor` creates abomination like 
`from_file(std::string filename, bool? shared=False, int64_t? size=c10::nullopt, *, ScalarType? dtype=c10::nullopt, ScalarType dtype=None, Layout layout=torch.strided, Device device=None, bool pin_memory=False, bool requires_grad=False)`

Im thinking if we can update options parsing to fold any subset of dtype, layout, device, .. whatever.. into TensorOptions, but also check if something unexpected passed (however this is out of scope for this PR)

**Comment by SsnL** [comment]
Will the python api eventually support loading from a file-like object?

File: aten/src/ATen/native/native_functions.yaml [hunk 5]
```diff
@@ -845,6 +845,10 @@
 
 - func: full_like(Tensor self, Scalar fill_value, *, ScalarType dtype, Layout layout, Device device, bool pin_memory=False) -> Tensor
 
+- func: from_file(str filename, bool? shared=None, int? size=0, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
```
Comment (apaszke) [inline aten/src/ATen/native/native_functions.yaml]: Why do we need `pin_memory` to be an optional bool? Isn't the default simply `false`?

Comment (VitalyFedyunin) [reply aten/src/ATen/native/native_functions.yaml]: Because our codegen folds TensorOptions only if they all specified similarly.

Comment (apaszke) [reply aten/src/ATen/native/native_functions.yaml]: In that case why did we miss the tracer functions prior to that patch?

**Pull request closed without merging**